kml/serdes.cpp

```diff
@@ -455,43 +455,106 @@ void SaveTrackLayer(Writer & writer, TrackLayer const & layer,
   {
-    if (e.empty())
+    auto const & line = geom.m_lines[lineIndex];
+    if (line.empty())
     {
-      ASSERT(false, ());
+      ASSERT(false, ("Unexpected empty Track"));
       continue;
     }
 
+    // Skip the tracks with the timestamps when writing the <LineString> points.
+    if (geom.HasTimestampsFor(lineIndex))
+      continue;
+
     writer << linesIndent << "<LineString><coordinates>";
 
-    writer << PointToString(e[0]);
-    for (size_t i = 1; i < e.size(); ++i)
-      writer << " " << PointToString(e[i]);
+    writer << PointToString(line[0]);
+    for (size_t pointIndex = 1; pointIndex < line.size(); ++pointIndex)
+      writer << " " << PointToString(line[pointIndex]);
 
     writer << "</coordinates></LineString>\n";
   }
 
-  if (sz > 1)
+  if (lineStringsSize > 1)
     writer << kIndent4 << "</MultiGeometry>\n";
 }
 
+void SaveGxTracks(Writer & writer, MultiGeometry const & geom)
+{
+  auto linesIndent = kIndent4;
+  auto const gxTracksSize = geom.GetNumberOfLinesWithTimestamps();
+  
+  if (gxTracksSize > 1)
+  {
+    linesIndent = kIndent8;
+    writer << kIndent4 << "<gx:MultiTrack>\n";
+    /// @TODO(KK): add the <altitudeMode>absolute</altitudeMode> if needed
+  }
+
+  for (size_t lineIndex = 0; lineIndex < geom.m_lines.size(); ++lineIndex)
+  {
+    auto const & line = geom.m_lines[lineIndex];
+    if (line.empty())
+    {
+      ASSERT(false, ("Unexpected empty Track"));
+      continue;
+    }
+
+    // Skip the tracks without the timestamps when writing the <gx:Track> points.
+    if (!geom.HasTimestampsFor(lineIndex))
+      continue;
+
+    writer << linesIndent << "<gx:Track>\n";
+    /// @TODO(KK): add the <altitudeMode>absolute</altitudeMode> if needed
+
+    auto const & timestampsForLine = geom.m_timestamps[lineIndex];
+    if (line.size() != timestampsForLine.size())
+      MYTHROW(KmlWriter::WriteKmlException, ("Timestamps size", timestampsForLine.size(),
+                                             "mismatch with the points size:", line.size(),
+                                             "for the track:", lineIndex));
+
+    for (size_t pointIndex = 0; pointIndex < line.size(); ++pointIndex)
+      writer << linesIndent << kIndent4 << "<when>" << base::SecondsSinceEpochToString(timestampsForLine[pointIndex]) << "</when>\n";
+
+    for (const auto & point : line)
+      writer << linesIndent << kIndent4 << "<gx:coord>" << PointToGxString(point) << "</gx:coord>\n";
+
+    writer << linesIndent << "</gx:Track>\n";
+  }
+
+  if (gxTracksSize > 1)
+    writer << kIndent4 << "</gx:MultiTrack>\n";
+}
+
+void SaveTrackGeometry(Writer & writer, MultiGeometry const & geom)
+{
+  size_t const sz = geom.m_lines.size();
+  if (sz == 0)
+  {
+    ASSERT(false, ("Unexpected empty MultiGeometry"));
+    return;
+  }
+
+  if (geom.m_lines.size() != geom.m_timestamps.size())
+    CHECK_EQUAL(geom.m_lines.size(), geom.m_timestamps.size(), ("Number of coordinates and timestamps should match"));
```

Comment:
Why if? CHECK_EQUAL already compares two values.
```suggestion
  CHECK_EQUAL(geom.m_lines.size(), geom.m_timestamps.size(), ("Number of coordinates and timestamps should match"));
```